gson/src/main/java/com/google/gson/internal/bind/JsogRegistry.java

```diff
@@ -0,0 +1,142 @@
+    private JsogPolicy jsogPolicy = GsonLocal.gson().getJsogPolicy();
+    /** Maps ids to instances */
+    private Map<String, Object> idToInstance = new HashMap<String, Object>();
+    /** Maps instances to ids */
+    private Map<Object, String> instanceToId = new IdentityHashMap<Object, String>();
+    /** The next JSOG id that should be assigned to each prefix */
+    private Map<String, Integer> nextIdByPrefix = new HashMap<String, Integer>();
+
+    /** Cannot get an instance directly, only via {@link #get()} */
+    private JsogRegistry() {}
+
+    /**
+     * Returns the JSOG registry associated with the current serialization / deserialization.
+     */
+    public static JsogRegistry get() {
+        return currentRegistry.get();
+    }
+
+    /**
+     * Returns the JSOG id associated with this instance, or null
+     * if the instance has not been registered
+     */
+    public String geId(Object instance) {
+        return instanceToId.get(instance);
+    }
+
+    /**
+     * Returns the instance associated with the given JSOG id, or null
+     * if the id has not been registered
+     */
+    public Object getInstance(String id) {
+        return idToInstance.get(id);
+    }
+
+    /**
+     * Register an instance with an auto-generated id
+     *
+     * @param instance Instance being registered
+     * @return The JSOG id associated with the instance
+     * @throws IllegalStateException If the object has already been registered
+     * @see JsogPolicy#getIdPrefix(Object)
+     */
+    public String register(Object instance) {
+        String prefix = jsogPolicy.getIdPrefix(instance);
+        if (prefix == null) {
+            prefix = "";
+        } else {
+            prefix += "-";
+        }
+
+        Integer id = nextIdByPrefix.get(prefix);
+        if (id == null) {
+            id = 1;
+        }
+
+        // Ensure automatic IDs don't conflict with manually-registered IDs
+        while (idToInstance.containsKey(prefix + id)) {
+            id++;
+        }
+        String fullId = prefix + id;
+        register(fullId, instance);
+        nextIdByPrefix.put(prefix, id + 1);
+        return fullId;
+    }
+
+    /**
+     * Register an instance with an auto-generated JSOG id
+     *
+     * @param id JSOG id of the instance
+     * @param instance Instance being registered
+     * @throws IllegalStateException If the object or the id have already been registered
+     */
+    public void register(String id, Object instance) {
+        if (id == null) {
+            throw new NullPointerException("null id");
+        }
+        if (instance == null) {
+            throw new NullPointerException("null instance");
+        }
+        if (idToInstance.containsKey(id)) {
+            throw new IllegalStateException("JsogEnabled id already registered: " + id);
+        }
+        if (instanceToId.containsKey(instance)) {
+            throw new IllegalStateException("JsogEnabled instance already registered: " + instance);
```

Comment:
This might be incorrect because it can also be used without using the `@JsogEnabled` annotation, right?